modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/toolbarbuttons/plugin.js

```diff
@@ -50,22 +225,75 @@
 			});
 
 			editor.ui.addRichCombo('TableHeaders', {
+				_headersLabels: {
+					[HEADING_BOTH]: `${editor.lang.table.headers}: ${editor.lang.table.headersBoth}`,
+					[HEADING_COL]: `${editor.lang.table.headers}: ${editor.lang.table.headersColumn}`,
+					[HEADING_NONE]: `${editor.lang.table.headers}: ${editor.lang.table.headersNone}`,
+					[HEADING_ROW]: `${editor.lang.table.headers}: ${editor.lang.table.headersRow}`,
+				},
+
+				_setHeading(heading, editor) {
+					setHeading({editor, heading, table: null});
+
+					this.setValue(heading, this._headersLabels[heading]);
+				},
+
 				init() {
-					const headersPrefix = editor.lang.table.headers;
+					this.add(
+						HEADING_NONE,
+						this._headersLabels[HEADING_NONE],
+						this._headersLabels[HEADING_NONE]
+					);
+					this.add(
+						HEADING_ROW,
+						this._headersLabels[HEADING_ROW],
+						this._headersLabels[HEADING_ROW]
+					);
+					this.add(
+						HEADING_COL,
+						this._headersLabels[HEADING_COL],
+						this._headersLabels[HEADING_COL]
+					);
+					this.add(
+						HEADING_BOTH,
+						this._headersLabels[HEADING_BOTH],
+						this._headersLabels[HEADING_BOTH]
+					);
 
-					const headersNone = `${headersPrefix}: ${editor.lang.table.headersNone}`;
-					const headersRow = `${headersPrefix}: ${editor.lang.table.headersRow}`;
-					const headersColumn = `${headersPrefix}: ${editor.lang.table.headersColumn}`;
-					const headersBoth = `${headersPrefix}: ${editor.lang.table.headersBoth}`;
+					const currentHeading = getHeading({editor, table: null});
 
-					this.add(headersNone, headersNone, headersNone);
-					this.add(headersRow, headersRow, headersRow);
-					this.add(headersColumn, headersColumn, headersColumn);
-					this.add(headersBoth, headersBoth, headersBoth);
+					this._setHeading(currentHeading, editor);
 				},
 
 				label: editor.lang.table.headers,
 
+				onClick(selectedHeading) {
+					this._setHeading(selectedHeading, editor);
+				},
+
+				onRender() {
+					const currentHeading = getHeading({editor, table: null});
+
+					this._setHeading(currentHeading, editor);
+
+					editor.on(
+						'selectionChange',
+						function (event) {
+							const editor = event.editor;
+
+							if (getFromSelection(editor)) {
+								const currentHeading = getHeading({
+									editor,
+									table: null,
+								});
+
+								this._setHeading(currentHeading, editor);
+							}
+						},
+						this
+					);
+				},
```

Comment:
The issue here is that `onRender` runs **before** rendering, you can see the relevant code in [richcombo](https://github.com/ckeditor/ckeditor4/blob/release/4.17.x/plugins/richcombo/plugin.js#L226-L243) and [balloontoolbar](https://github.com/ckeditor/ckeditor4/blob/release/4.17.x/plugins/balloontoolbar/plugin.js#L793-L801). The timeline is:
1. HTML is created from template and set into `output`.
2. `onRender` is invoked.
3. `setHtml` is invoked, based on `output`.

The first two lines in this block of code (L275-277) end up doing nothing. We set heading, and right after it resets to template value from `output`. If you set breakpoint on `setHtml`, you'll see the value updated before re-render starts.

`selectionChange` is what does the job of updating value, but it's a very ugly patch. We are using a selection, an event that happens to be at the right time, to have an effect completely unrelated to selection. This type of unrelated wiring may cause very weird and hard-to-find bugs in the future.

Instead, what we can do is update template params, before template is set:
```js
onRender() {
	const currentHeading = getCurrentHeading({editor});

	this.label = this._headersLabels[currentHeading];
},
```